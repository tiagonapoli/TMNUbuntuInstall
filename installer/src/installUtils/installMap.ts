
export interface InstallMapNode {
  name: string
  children?: InstallMapNode[]
  scriptPath?: string
  configSubPath?: string
}

export interface InstallStep {
  name: string
  scriptPath: string
  config?: any
}

export function generateInstallSteps(
  node: InstallMapNode,
  config: any,
  configPath = '',
  treePath = ''
): InstallStep[] {
  const { name, scriptPath, children, configSubPath } = node

  const resultTreePath = treePath ? `${treePath}:${name}` : name
  const resultConfigPath = configPath ? `${configPath}.${configSubPath}` : configSubPath

  if (configSubPath && !config?.[configSubPath]) {
    throw new Error(`Missing config for ${resultTreePath}. Needs ${resultConfigPath} key on config.`)
  }

  if (!children || children.length === 0) {
    return [
      {
        name: resultTreePath,
        scriptPath,
        ...(configSubPath ? { config: config[configSubPath] } : null),
      },
    ]
  }

  let res: InstallStep[] = []
  children.forEach(childNode => {
    res.push(
      ...generateInstallSteps(childNode, configSubPath ? config[configSubPath] : null, resultConfigPath, resultTreePath)
    )
  })
  return res
}
