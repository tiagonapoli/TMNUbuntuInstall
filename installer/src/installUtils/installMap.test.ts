import { generateInstallSteps, InstallMapNode, InstallStep } from './installMap'

interface TreeSample {
  tree: InstallMapNode
  config: any
}

interface ErrorSample extends TreeSample {
  throws: true
}

interface SuccessSample extends TreeSample {
  res: InstallStep[]
}

const samples: Array<ErrorSample | SuccessSample> = [
  {
    tree: {
      name: 'a',
      scriptPath: 'a.sh',
      configSubPath: 'a',
      children: [
        { name: 'b', scriptPath: 'b.sh', configSubPath: 'b' },
        { name: 'c', scriptPath: 'c.sh' },
      ],
    },
    config: {
      a: {
        b: 'b',
        c: 'c',
      },
    },
    res: [
      { name: 'a:b', scriptPath: 'b.sh', config: 'b' },
      { name: 'a:c', scriptPath: 'c.sh' },
    ],
  },

  {
    throws: true,
    tree: {
      name: 'a',
      scriptPath: 'a.sh',
      configSubPath: 'a',
      children: [{ name: 'b', scriptPath: 'b.sh', configSubPath: 'b' }],
    },
    config: {
      a: {},
    },
  },
]

test.each(samples)('Test %#', sample => {
  const { config, tree } = sample
  if ((sample as ErrorSample).throws) {
    expect(() => generateInstallSteps(tree, config)).toThrowErrorMatchingSnapshot()
  } else {
    expect(generateInstallSteps(tree, config)).toEqual((sample as SuccessSample).res)
  }
})
