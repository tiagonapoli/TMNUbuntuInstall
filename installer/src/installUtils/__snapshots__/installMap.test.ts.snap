// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test 1 1`] = `"Missing config for a:b. Needs a.b key on config."`;
